File: LeetCode/200NumberOfIslands/NumberOfIslands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        # DFS -> Time:O(nm), Space:O(nm)
        ans = 0
        visited = set()

        def dfs(r, c):
            if r >= 0 and \
                c >= 0 and \
                r < len(grid) and \
                c < len(grid[0]) and \
                (r, c) not in visited and \
                grid[r][c] == '1':
                visited.add((r, c))
                dfs(r+1, c)
                dfs(r, c+1)
                dfs(r-1, c)
                dfs(r, c-1)

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1' and (i, j) not in visited:
                    ans += 1
                    dfs(i, j)

        return ans

        # # BFS -> Time:O(nm), Space:O(nm)
        # if not grid:
        #     return 0
        # visited = set()
        # ans = 0
        # def bfs(r,c):
        #     q = []
        #     q.append((r,c))
        #     while q:
        #         row,col = q.pop(0)
        #         directions = [[-1,0], [1,0], [0, 1], [0, -1]]
        #         for dr,dc in directions:
        #             r, c = row+dr, col+dc
        #             if (r in range(len(grid)) and
        #                 c in range(len(grid[0])) and
        #                 grid[r][c] == "1" and 
        #                 (r,c) not in visited):
        #                 q.append((r,c))
        #                 visited.add((r,c))
        # for r in range(len(grid)):
        #     for c in range(len(grid[0])):
        #         if (r,c) not in visited and grid[r][c] == "1":
        #             bfs(r,c)
        #             visited.add((r,c))
        #             ans += 1
        # return ans

File: LeetCode/200NumberOfIslands/NumberOfIslands.cpp
class Solution {
public:
    int ans = 0;
    set<vector<int>> visited;
    vector<vector<int>> directions {
        {1, 0},
        {-1, 0},
        {0, 1},
        {0, -1}
    };
    void bfs(vector<vector<char>>& grid, vector<int> v) {
        queue<vector<int>> q;
        q.push(v);
        while (!q.empty()) {
            vector<int> current = q.front();
            q.pop();
            for (int i = 0; i < 4; i++) {
                int row = directions[i][0] + current[0];
                int col = directions[i][1] + current[1];
                vector<int> temp;
                temp.push_back(row);
                temp.push_back(col);
                if (row >= 0 &&
                    row < grid.size() &&
                    col >= 0 &&
                    col < grid[0].size() &&
                    visited.find(temp) == visited.end() &&
                    grid[row][col] == '1') {
                        q.push(temp);
                        visited.insert(temp);
                    }
            }
        }
    }
    int numIslands(vector<vector<char>>& grid) {
        // BFS -> Time:O(nm), Space:O(nm)
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++) {
                vector<int> v;
                v.push_back(i);
                v.push_back(j);
                if (grid[i][j] == '1' && visited.find(v) == visited.end()) {
                    bfs(grid, v);
                    visited.insert(v);
                    ans++;
                }
            }
        }
        return ans;
    }
};
